src/core/src/core_action_controller.cpp

```diff
@@ -434,90 +444,118 @@ bool CoreActionController::saveSong() {
 	auto pSong = pHydrogen->getSong();
 	
 	// Extract the path to the associate .h2song file.
-	QString songPath = pSong->get_filename();
+	QString sSongPath = pSong->get_filename();
 	
-	if ( songPath.isEmpty() ) {
+	if ( sSongPath.isEmpty() ) {
 		ERRORLOG( "Unable to save song. Empty filename!" );
 		return false;
 	}
 	
 	// Actual saving
-	bool saved = pSong->save( songPath );
+	bool saved = pSong->save( sSongPath );
 	if ( !saved ) {
 		ERRORLOG( QString( "Current song [%1] could not be saved!" )
-				  .arg( songPath ) );
+				  .arg( sSongPath ) );
 		return false;
 	}
 	
 	// Update the status bar.
-	if ( pHydrogen->getActiveGUI() ) {
-		EventQueue::get_instance()->push_event( EVENT_UPDATE_SONG, 1 );
+	if ( pHydrogen->getGUIState() != Hydrogen::GUIState::unavailable ) {
+		EventQueue::get_instance()->push_event( EVENT_UPDATE_SONG, 2 );
 	}
 	
 	return true;
 }
 
-bool CoreActionController::saveSongAs( const QString& songPath ) {
+bool CoreActionController::saveSongAs( const QString& sSongPath ) {
 	
 	auto pHydrogen = Hydrogen::get_instance();
 	
 	// Get the current Song which is about to be saved.
 	auto pSong = pHydrogen->getSong();
 	
 	// Check whether the provided path is valid.
-	if ( !isSongPathValid( songPath ) ) {
+	if ( !isSongPathValid( sSongPath ) ) {
 		// isSongPathValid takes care of the error log message.
 		return false;
 	}
 	
-	if ( songPath.isEmpty() ) {
+	if ( sSongPath.isEmpty() ) {
 		ERRORLOG( "Unable to save song. Empty filename!" );
 		return false;
 	}
 	
 	// Actual saving
-	bool saved = pSong->save( songPath );
-	if ( !saved ) {
+	bool bSaved = pSong->save( sSongPath );
+	if ( !bSaved ) {
 		ERRORLOG( QString( "Current song [%1] could not be saved!" )
-				  .arg( songPath ) );
+				  .arg( sSongPath ) );
 		return false;
 	}
 	
 	// Update the status bar.
-	if ( pHydrogen->getActiveGUI() ) {
-		EventQueue::get_instance()->push_event( EVENT_UPDATE_SONG, 1 );
+	if ( pHydrogen->getGUIState() != Hydrogen::GUIState::unavailable ) {
+		EventQueue::get_instance()->push_event( EVENT_UPDATE_SONG, 2 );
 	}
 	
 	return true;
 }
 
+bool CoreActionController::savePreferences() {
+	
+	if ( Hydrogen::get_instance()->getGUIState() != Hydrogen::GUIState::unavailable ) {
+		// Update the status bar and let the GUI save the preferences
+		// (after writing its current settings to disk).
+		EventQueue::get_instance()->push_event( EVENT_UPDATE_PREFERENCES, 0 );
+	} else {
+		Preferences::get_instance()->savePreferences();
+	}
+	
+	return true;
+
+}
+
+bool CoreActionController::saveDrumkit() {
+
+	return true;
```

Comment:
This looks like a stub :) Is this intended?

Reply:
You are right. I decided against implementing this one :)